File: leetcode-js/163-missing-ranges.js
/**
 * @param {number[]} nums
 * @param {number} lower
 * @param {number} upper
 * @return {string[]}
 */
var findMissingRanges = function(nums, lower, upper) {
    const result = [];

    let pre = lower - 1;

    for (let i = 0; i <= nums.length; i += 1) {
        const after = i === nums.length ? upper + 1 : nums[i];
        if (pre + 2 === after) {
            result.push(`${pre + 1}`);
        } else if (pre + 2 < after) {
            result.push(`${pre + 1}->${after - 1}`);
        }
        pre = after;
    }

    return result;
};

File: leetcode-java/163-missing-ranges.java
public class Solution {
    public List<String> findMissingRanges(int[] nums, int lower, int upper) {
        List<String> res = new ArrayList<>();
        int pre = lower - 1;
        for (int i = 0; i <= nums.length; i++) {
            int n = i == nums.length ? upper + 1 : nums[i];
            if (pre + 2 == n) {
                res.add(String.valueOf(pre + 1));
            } else if (pre + 2 < n) {
                res.add(String.valueOf(pre + 1) + "->" + String.valueOf(n - 1));
            }
            pre = n;
        }
        return res;
    }
}
